### PR TITLE
Check interpolated string arguments in Binder.GetValEscape()

### LINKED ISSUE #63262: Unexpected escape analysis behavior for interpolated string with variable of local scope
The escape analysis behavior for the following interpolated string cases is unexpected:
```csharp
using System;
using System.Runtime.CompilerServices;

[InterpolatedStringHandler]
ref struct CustomHandler
{
    public CustomHandler(int literalLength, int formattedCount) { }
    public void AppendFormatted(Span<char> s) { }
}

class Program
{
    static CustomHandler F1()
    {
        Span<char> s = stackalloc char[10];
        CustomHandler h1 = $"{s}";
        return h1; // no error
    }

    static void F2()
    {
        Span<char> s = stackalloc char[10];
        F3($"{s}"); // error CS8352: ... may expose variables outside declaration scope
    }

    static void F3(ref CustomHandler handler) { }
}
```

For `F1()`, no error is reported returning `h1` although it may capture the local `Span<char>` from `$"{s}"`.

For `F2()`, an error is reported calling `F3($"{s}")`, although it seems no references should escape `F3()`.

### PR DESCRIPTION
**Pull request by cston**
Fixes #63262

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 1]
```diff
@@ -13815,6 +13815,80 @@ static void Main()
...
+        public void RefEscape_12(LanguageVersion languageVersion)
+        {
+            var code =
+@"using System;
+using System.Runtime.CompilerServices;
+[InterpolatedStringHandler]
+ref struct CustomHandler
+{
+    public CustomHandler(int literalLength, int formattedCount) { }
+    public void AppendFormatted(Span<char> s) { }
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:13860]: Please add a version of this test with a `scoped` parameter. /cc @stephentoub, presumably `DefaultInterpolatedStringHandler` will need to react to this breaking change.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 1). /cc @stephentoub, presumably `DefaultInterpolatedStringHandler` will need to react to this breaking change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 2]
```diff
@@ -13815,6 +13815,80 @@ static void Main()
...
+}
+class Program
+{
+    static void Main()
+    {
+        Span<char> s = stackalloc char[10];
+        M($""{s}"");
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:13838]: consider also including the version of this scenario which explicitly calls Appendformatted, etc. similar to RefEscape_12.

**Review by RikkiGibson — APPROVED** [review]
LGTM modulo the review comments.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 3]
```diff
@@ -13832,23 +13832,71 @@ public void RefEscape_11(LanguageVersion languageVersion)
...
+    static void F1()
     {
         Span<char> s = stackalloc char[10];
         M($""{s}"");
     }
+    static void F2()
+    {
+        Span<char> s = stackalloc char[10];
+        CustomHandler h2 = new CustomHandler(0, 1);
+        h2.AppendFormatted(s); // 1
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:13844]: I didn't understand why the interpolated string version had no errors but the "explicit" version has this error. Could you please clarify?

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:13844]: My understanding is the interpolated string version is considered as creating a `CustomHandler` instance as if the `new CustomHandler()` and the interpolated string arguments were combined as an object initializer. In `F1()`, where the interpolated string argument `s` has safe-to-escape of the current method, the resulting `CustomHandler` instance will have safe-to-escape of the current method as well. In `F2()`, the argument `s` is not included in the `CustomHandler` object initializer, so the resulting instance is safe-to-escape to the caller.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:13844]: > My understanding is the interpolated string version is considered as creating a CustomHandler instance as if the new CustomHandler() and the interpolated string arguments were combined as an object initializer.

I would expect these scenarios to have the same behavior.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 4]
```diff
@@ -13865,30 +13913,78 @@ static CustomHandler F1()
...
+    public CustomHandler(int literalLength, int formattedCount) { }
+    public void AppendFormatted(scoped Span<char> s) { }
+}
+class Program
+{
+    static CustomHandler F1()
+    {
+        Span<char> s = stackalloc char[10];
+        CustomHandler h1 = $""{s}"";
+        return h1; // 1
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:13960]: Why is this occurring? The `AppendFormatted` call is marked `scoped`, so the `safe-to-escape` of `h1` should not be restricted to this method. This _will_ break the BCL and other libraries if it is merged as is.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 2). The current behavior is not what I'd expect from the feature.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs (hunk 4 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs]: Good catch, thanks.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:13873]: > I would expect these scenarios to have the same behavior.

If we make the `CustomHandler` in the first example explicit, the two examples are essentially:
```csharp
CustomHandler h1 = $"{s}";
M(h1);

CustomHandler h2 = new CustomHandler(0, 1);
h2.AppendFormatted(s);
M(h2);
```

The difference between the two is the initializers: the initializer for `h1` limits the scope to the current method; the initializer for `h2` does not limit the scope.

**Review by 333fred — APPROVED** [review]

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**